### ISSUE TITLE
timeout

### ISSUE THREAD

**Issue by amarco** [half2me/libant#8]
I am getting the following when running on windows.
usb.core.USBError: [Errno None] b'libusb0-dll:err [_usb_reap_async] timeout error\n'

Any idea where I can look for this?

**Comment by half2me** [comment]
Hi @amarco its been awhile since I've worked on this. If at all possible, I would suggest using my [antgo](https://github.com/half2me/antgo) repo instead.

Are you sure you are loading the libusb driver instead of winUSB? You can use this tool to easily switch between the two: https://zadig.akeo.ie/

**Comment by amarco** [comment]
Hi Thanks for the reply.
Unfortunately my goal is to integrate the code into an existing python application so I am not sure the go approach would work for me.
This is why I am starting with the examples/demos to ensure I can integrate it.

Correct I am using the zadiag tool to load the libusb instead of winusb. Oddly enough I experimented with zadiag and the problems persists across different drivers.

I am fairly new to python and looking at your code example I see you used “Factory” which is a construct I am not familiar with but rather looks cool.
What I am wondering how I could change things to have the cadence and speed be class properties so that in the top callback I can access the data as variables.
I have been looking at code and am not sure how to go about adding this feature.

In other words

from time import sleep

from libAnt.drivers.usb import USBDriver
from libAnt.loggers.pcap import PcapLogger
from libAnt.node import Node
from libAnt.profiles.factory import Factory


def callback(msg):
<here I would like to access speed and cadence as variables instead of sensing the message to print>
       I tried msg.cadence but it was always zero.
    print(msg)


def eCallback(e):
    raise (e)


f = Factory(callback)

stick =USBDriver(vid=0x0FCF, pid=0x1008)
ant=Node(stick,'RiderNet')

ant.enableRxScanMode()
ant.start(f.parseMessage, eCallback)

sleep(10)
ant.stop()
input('Press enter to continue: ')

**Comment by amarco** [comment]
Hi Again

So this seems to work for me…

from time import sleep

from libAnt.drivers.usb import USBDriver
from libAnt.loggers.pcap import PcapLogger
from libAnt.node import Node
from libAnt.profiles.factory import Factory


def callback(msg):
    #print(msg)
    print(msg['Cadence'])


def eCallback(e):
    raise (e)


f = Factory(callback)

stick =USBDriver(vid=0x0FCF, pid=0x1008)
ant=Node(stick,'RiderNet')

ant.stop()
sleep(1)
ant.enableRxScanMode()
ant.start(f.parseMessage, eCallback)

and in the factory class I changed the parseMessage method to call a new method in SpeedAndCadenceProfileMessage class
    def parseMessage(self, msg: BroadcastMessage):
        with self._lock:
            if self._filter is not None:
                if msg.deviceNumber not in self._filter:
                    return
            if msg.deviceType in Factory.types:
                num = msg.deviceNumber
                type = msg.deviceType
                if type == 11: # Quick patch to filter out power messages with non-power info
                    if msg.content[0] != 16:
                        return
                pmsg = self.types[type](msg, self._messages[(num, type)] if (num, type) in self._messages else None)
                self._messages[(num, type)] = pmsg
                #if callable(self._callback):
                #    self._callback(pmsg)
                data = self.types[type].getResults(pmsg)
                if callable(self._callback):
                    self._callback(data)


the new method in SpeedAndCadenceProfileMessage class looks as follows
    def getResults(self):
        #print('getting results')
        results={}
        results={'Speed': self.speed(2096) ,
                 'avgSpeed':self.averageSpeed(2096),
                 'Cadence':self.cadence,
                 'avgCadence': self.averageCadence,
                 'totalDistance':self.totalDistance(2096),
                 'totalRevolutions': self.totalRevolutions}
        return results

It seems to give me the behavior I want, would you of done it this way?

Marco

**Comment by amarco** [comment]
Well I am back at this project and still have the timeout problem.
I noticed I did not respond to your question about what driver is loaded and yes it is the libusb0 (v1.2.6.0) as indicated by zadig.
The crazy thing is that it works sometime and does not other times and when it gets in a funk it stays that way. I did a USB capture but I am not able to tell anything from it other than the pipe is closed.
Just not sure where to look, forest for the trees right now.
Marco

**Comment by half2me** [comment]
hey @amarco sorry, its been ages since I've done anything with this repo. Make sure you are using the correct: pid=0x1008. (sometimes its 0x1009 depending on hw). However, if you say it works sometimes, could there be any other app who is already using the libusb resource?

**Comment by amarco** [comment]
Hi half2me, thanks for the response.
Got some progress yesterday after playing with different python backends.
Yes the pid needs to be 0x1009 for the Ant USB-m Stick.
I am also using the newer Speed 2 and Cadence 2 sensor since GSC10 all in one is no longer available.
Looking at your factories I did not see a separate profile for a speed sensor only and a cadence sensor only. 
I did notice that the two devices  produce a type id of  122 for the Cadence only sensor and  123 for the Speed only sensor.
Do you have any profiles for speed only and cadence only sensors?
Is it a matter of using the combined profile and adjusting array offsets?

**Comment by half2me** [comment]
@amarco No I don't have profiles for those, but you can create them them, and submit a PR if you would like :) You just need to use the correct offsets. All the info on how to decode this information should be available on the [ANT+ Dev Site](https://www.thisisant.com/developer/resources/downloads/#documents_tab) if you are logged in. They have a pdf for each profile.

**Comment by amarco** [comment]
Thanks, I found the pdfs and was able to create two new profiles with offsets adjusted ect. They seem to work and provide numbers that are not wild :)
Also what do you mean by submit a PR?

**Comment by amarco** [comment]
Just an fyi on my windows 10 box the libusb1.0 backend seems to work much better for me compared to libusb0.1 while using the native Ant USB-m stick drivers. So good progress in the last 48 hours :)

**Comment by half2me** [comment]
> Thanks, I found the pdfs and was able to create two new profiles with offsets adjusted ect. They seem to work and provide numbers that are not wild :)
> Also what do you mean by submit a PR?

I mean you can submit a Pull Request here on github, with the profiles that you created. That way others can use them as well. :)

**Comment by half2me** [comment]
> Just an fyi on my windows 10 box the libusb1.0 backend seems to work much better for me compared to libusb0.1 while using the native Ant USB-m stick drivers. So good progress in the last 48 hours :)

Glad to hear that. Good luck :)
